crates/rattler_installs_packages/src/python_env/byte_code_compiler.rs

```diff
@@ -0,0 +1,279 @@
+use parking_lot::Mutex;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::{
+    io::{self, BufRead, BufReader, Write},
+    path::{Path, PathBuf},
+    process::{Command, Stdio},
+    sync::mpsc::{channel, Sender},
+};
+use tempfile::NamedTempFile;
+use thiserror::Error;
+
+/// The result of compiling a Python script to byte code.
+type CompilationResponse = Result<PathBuf, CompilationError>;
+type CompilationRequest = PathBuf;
+
+type BoxedCallback = Box<dyn FnOnce(CompilationResponse) + Send + 'static>;
+
+/// An error that can occur when compiling a source file.
+#[derive(Debug, Error, Clone)]
+pub enum CompilationError {
+    /// The file is not a python file.
+    #[error("not a python file")]
+    NotAPythonFile,
+
+    /// The file could not be found.
+    #[error("source file not found")]
+    SourceNotFound,
+
+    /// The file could not be compiled
+    #[error("failed to compile")]
+    FailedToCompile,
+
+    /// The compilation host unexpectedly quit.
+    #[error("host has quit")]
+    HostQuit,
+}
+
+/// An error that can occur when spawning the compilation host.
+#[derive(Debug, Error)]
+pub enum SpawnCompilerError {
+    /// Could not create the source code that runs in the compilation host.
+    #[error("failed to create temporary file for compilation source")]
+    FailedToCreateSource(#[source] io::Error),
+
+    /// Failed to start the python executable
+    #[error("failed to start python executable")]
+    FailedToStartPython(#[source] io::Error),
+}
+
+/// An object that allows compiling python source code to byte code in a separate process.
+pub struct ByteCodeCompiler {
+    /// The channel that is used to send compilation requests to the compilation host. If this is
+    /// dropped the attached thread will drop stdin of the child which will signal the child to
+    /// exit.
+    request_tx: Option<Sender<CompilationRequest>>,
+
+    /// Callback functions per compilation request. These are called when the compilation host
+    /// finishes processing a request.
+    pending_callbacks: Arc<Mutex<HashMap<CompilationRequest, Vec<BoxedCallback>>>>,
+
+    /// The child process. This is waited upon on drop.
+    child: Option<std::process::Child>,
+
+    // The file that contains the python source code of the compilation host. We keep this around
+    // to make sure the file is not accidentally deleted while the compilation host is still using
+    // it.
```

Comment:
I think this is actually only potentially a problem on Windows, right? On macOS and Linux it shouldn't matter if the file is deleted while Python is running.

Reply:
We need to make sure the file exists when python is still initializing. The `new` method starts the executable but there is no way to be certain that python already read the source code. To that end, I just keep the file around as long as the process is alive but in theory it could also just be deleted somewhere earlier.